src/jit/MemoryInl.h

```diff
@@ -1742,118 +1742,139 @@ static void emitAtomic(sljit_compiler* compiler, Instruction* instr)
-        sljit_emit_op1(compiler, operationSize, tmpReg, 0, srcValue.arg, srcValue.argw);
+    struct sljit_label* restartOnFailure = sljit_emit_label(compiler);
+
+    if (noShortAtomic && size <= 2) {
+        sljit_emit_op1(compiler, SLJIT_MOV32, maskReg, 0, SLJIT_MEM1(SLJIT_SP), context->stackTmpStart);
+        sljit_emit_op2(compiler, SLJIT_SHL32, maskReg, 0, SLJIT_IMM, size == 1 ? 0xff : 0xffff, maskReg, 0);
     }
 
     sljit_emit_atomic_load(compiler, operationSize, SLJIT_TMP_DEST_REG, baseReg);
 
     if (noShortAtomic && size <= 2) {
-        sljit_emit_op2(compiler, SLJIT_AND32, tmpReg, 0, SLJIT_TMP_DEST_REG, 0, maskReg, 0);
-        sljit_emit_op2(compiler, SLJIT_LSHR32, tmpReg, 0, tmpReg, 0, shiftReg, 0);
+        sljit_emit_op2(compiler, SLJIT_AND32, maskReg, 0, maskReg, 0, SLJIT_TMP_DEST_REG, 0);
     }
 
-    compareFalse = sljit_emit_cmp(compiler, SLJIT_NOT_EQUAL, (noShortAtomic && size <= 2 ? tmpReg : SLJIT_TMP_DEST_REG), 0, srcExpectedReg, 0);
+    sljit_s32 compare = SLJIT_NOT_EQUAL | (operationSize & SLJIT_32) | (operationSize == SLJIT_MOV32 ? SLJIT_32 : 0);
+    struct sljit_jump* compareFalse = sljit_emit_cmp(compiler, compare, maskReg, 0, srcExpectedReg, 0);
 
     if (noShortAtomic && size <= 2) {
-        sljit_emit_op2(compiler, SLJIT_SHL32, tmpReg, 0, tmpReg, 0, shiftReg, 0);
+        sljit_emit_op2(compiler, SLJIT_XOR32, maskReg, 0, maskReg, 0, srcValueReg, 0);
+        sljit_emit_op2(compiler, SLJIT_XOR32, maskReg, 0, maskReg, 0, SLJIT_TMP_DEST_REG, 0);
+        srcValueReg = maskReg;
+    }
 
-        sljit_emit_op1(compiler, SLJIT_MOV32, srcExpectedReg, 0, SLJIT_TMP_DEST_REG, 0);
-        sljit_emit_op2(compiler, SLJIT_XOR32, maskReg, 0, maskReg, 0, SLJIT_IMM, -1);
-        sljit_emit_op2(compiler, SLJIT_AND32, srcExpectedReg, 0, srcExpectedReg, 0, maskReg, 0);
-        sljit_emit_op2(compiler, SLJIT_OR32, tmpReg, 0, srcValueReg, 0, srcExpectedReg, 0);
+    sljit_emit_atomic_store(compiler, operationSize | SLJIT_SET_ATOMIC_STORED, srcValueReg, baseReg, SLJIT_TMP_DEST_REG);
+    sljit_set_label(sljit_emit_jump(compiler, SLJIT_ATOMIC_NOT_STORED), restartOnFailure);
+
+    if ((noShortAtomic && size <= 2) || (!is32BitOp && srcExpectedPair.arg2 != 0)) {
+        storeSuccess = sljit_emit_jump(compiler, SLJIT_ATOMIC_STORED);
     }
 
     if (noShortAtomic && size <= 2) {
-        sljit_emit_op1(compiler, SLJIT_MOV32, srcExpectedReg, 0, SLJIT_TMP_DEST_REG, 0);
+        sljit_set_label(compareTopFalse, sljit_emit_label(compiler));
     }
-    sljit_emit_atomic_store(compiler, operationSize | SLJIT_SET_ATOMIC_STORED, tmpReg, baseReg, SLJIT_TMP_DEST_REG);
-    if (noShortAtomic && size <= 2) {
-        sljit_emit_op1(compiler, SLJIT_MOV32, SLJIT_TMP_DEST_REG, 0, srcExpectedReg, 0);
+
+    if (!is32BitOp && srcExpectedPair.arg2 != 0) {
+        sljit_set_label(compareUpperHalfFalse, sljit_emit_label(compiler));
+        sljit_emit_op1(compiler, operationSize, SLJIT_TMP_DEST_REG, 0, SLJIT_MEM1(baseReg), 0);
```

Comment:
Does this always overwrite SLJIT_TMP_DEST_REG? This can be incorrect if another thread already changed the value.

Reply:
It did, if we returned early because the top half was not 0.
I changed it, so we check it after the atomic load, and use the atomically loaded value, instead of a separately loaded one.